### PR TITLE
Add in --merge-target option

### PR DESCRIPTION
**Pull request by alexlatchford**
At a high-level the `--merge-target` option adds a convenient mechanism for branching off the `HEAD` (usually `master`) of the project that's being updated. For example if you wanted to run `battenberg upgrade` on `pricing-triage` then it'd by default try to merge the changes back into `master`, this is fine but our flow is to then construct a PR that'd allow the Purchase ML team to approve the upgrade.

In practice this means you'd run: `battenberg upgrade --merge-target archetype-update`. You'd run create a PR from `archetype-update` -> `master` for the repo you're looking to upgrade (ie. `pricing-triage`).

Screen Shot 2019-10-07 at 3 10 06 PM

To generate this screenshot I ran:

```bash
cd /tmp
battenberg -O example-project install git@gitlab.zgtools.net:zillow-offers/zoml/ml-infrastructure/archetype.py-ml.git
cd example-project
# Add in the template context.
vi .cookiecutter.json
git add . && commit -m "Add cc.json"
battenberg upgrade --no-input --merge-target archetype-update --checkout archetypev1.1
```

Done:
- Resolves [ZOMLINFRA-406](https://zbrt.atl.zillow.net/browse/ZOMLINFRA-406).
- Also cleans up the usage docs and CLI interface some more.
- Deprecates the `show` command which is obsolete now we have the template context easily accessible via `.cookiecutter.json`.

### CODE REVIEW

File: battenberg/core.py [hunk 1]
```diff
@@ -77,27 +77,30 @@ def __exit__(self, type, value, traceback):
...
+        merge_target_ref = 'HEAD'
+        if merge_target is not None:
+            # If we have a merge target, ensure we have that branch and have switched to it
+            # before continuing with merging.
+            merge_target_ref = f'refs/heads/{merge_target}'
+            if merge_target not in self.repo.branches:
+                self.repo.branches.local.create(merge_target, self.repo.get(self.repo.head.target))
+            self.repo.checkout(merge_target_ref)
+
+        # Analyze merge between template branch and our merge target.
```
Comment (shoelsch) [inline battenberg/core.py]: Nit: Is this comment adding information?

File: battenberg/core.py [hunk 2]
```diff
@@ -186,21 +190,24 @@ def install(self, template, checkout='master', extra_context=None, no_input=Fals
         # Let's merge our changes into HEAD
         self.merge_template_branch(f'Installed template \'{template}\'')
 
-    def upgrade(self, checkout='master', extra_context=None, no_input=False):
+    def upgrade(self, checkout: str = 'master', extra_context: Dict = None, no_input: bool = False,
```
Comment (shoelsch) [inline battenberg/core.py]: Should we require user confirmation before proceeding with an upgrade? The default is usually still master, right?

File: battenberg/core.py [hunk 3]
```diff
@@ -243,4 +250,4 @@ def upgrade(self, checkout='master', extra_context=None, no_input=False):
         self.repo.lookup_branch(TEMPLATE_BRANCH).set_target(commit.hex)
 
         # Let's merge our changes into HEAD
-        self.merge_template_branch(f'Upgraded template \'{template}\'')
+        self.merge_template_branch(f'Upgraded template \'{template}\'', merge_target)
```
Comment (shoelsch) [inline battenberg/core.py]: Do we think there's value in exposing the commit message? Would it make more sense for us to control that internally?

File: battenberg/core.py (hunk 1 above)
Comment (alexlatchford) [reply battenberg/core.py]: Removed it

File: battenberg/core.py (hunk 2 above)
Comment (alexlatchford) [reply battenberg/core.py]: Added in https://zbrt.atl.zillow.net/browse/ZOMLINFRA-425 to cover this. Will add this if I get time in the sprint.

File: battenberg/core.py (hunk 3 above)
Comment (alexlatchford) [reply battenberg/core.py]: Makes sense to me to add this functionality. Added https://zbrt.atl.zillow.net/browse/ZOMLINFRA-426 to cover this.

File: README.md [hunk 4]
```diff
@@ -21,36 +21,28 @@ conflicts needed to be manually resolved for each upgrade merge. To minimize the
 
 ## Usage
 
-Install a [Cookiecutter](https://github.com/audreyr/cookiecutter)
-template on current directory:
+Install a [Cookiecutter](https://github.com/audreyr/cookiecutter) template:
```
Comment (sruthinagalla) [inline README.md]: shouldn't this link to our gitlab archetype?

Comment (alexlatchford) [reply README.md]: Nope this is a generic tool built on top of the `cookiecutter` project itself so it's actually unaware of any of Zillow's archetypes :D

**Pull request merged**